engine/googletest_result.cpp

```diff
@@ -0,0 +1,520 @@
+    else {
+        INV(status == "successful");
+        return engine::googletest_result(engine::googletest_result::successful);
+    }
+}
+
+
+/// Parses a test result that needs a reason.
+///
+/// \param status The result status name.
+/// \param rest The rest of the line after the status name.
+///
+/// \return An object representing the test result.
+///
+/// \throw format_error If the result is invalid (i.e. rest is invalid).
+///
+/// \pre status must be one of "broken", "disabled", "failed", or "skipped".
+static engine::googletest_result
+parse_with_reason(const std::string& status, const std::string& rest)
+{
+    using engine::googletest_result;
+
+    INV(!rest.empty());
+
+    if (status == "broken")
+        return googletest_result(googletest_result::broken, rest);
+    else if (status == "disabled")
+        return googletest_result(googletest_result::disabled, rest);
+    else if (status == "failed")
+        return googletest_result(googletest_result::failed, rest);
+    else if (status == "skipped")
+        return googletest_result(googletest_result::skipped, rest);
+    else
+        PRE_MSG(false, "Unexpected status");
+}
+
+
+/// Formats the termination status of a process to be used with validate_result.
+///
+/// \param status The status to format.
+///
+/// \return A string describing the status.
+static std::string
+format_status(const process::status& status)
+{
+    if (status.exited())
+        return F("exited with code %s") % status.exitstatus();
+    else if (status.signaled())
+        return F("received signal %s%s") % status.termsig() %
+            (status.coredump() ? " (core dumped)" : "");
+    else
+        return F("terminated in an unknown manner");
+}
+
+
+}  // anonymous namespace
+
+
+/// Constructs a raw result with a type.
+///
+/// The reason and the argument are left uninitialized.
+///
+/// \param type_ The type of the result.
+engine::googletest_result::googletest_result(const types type_) :
+    _type(type_)
+{
+}
+
+
+/// Constructs a raw result with a type and a reason.
+///
+/// The argument is left uninitialized.
+///
+/// \param type_ The type of the result.
+/// \param reason_ The reason for the result.
+engine::googletest_result::googletest_result(const types type_,
+                               const std::string& reason_) :
+    _type(type_), _reason(reason_)
+{
+}
+
+
+/// Constructs a raw result with a type, an optional argument and a reason.
+///
+/// \param type_ The type of the result.
+/// \param argument_ The optional argument for the result.
+/// \param reason_ The reason for the result.
+engine::googletest_result::googletest_result(const types type_,
+                               const utils::optional< int >& argument_,
+                               const std::string& reason_) :
+    _type(type_), _argument(argument_), _reason(reason_)
+{
+}
+
+
+const std::string invalid_output_message = "invalid output";
```

Comment:
These constants are neither in an anonymous namespace nor static, so they'll cause trouble. Move to the beginning of the file.